src/cryptonote_core/service_node_voting.cpp

```diff
@@ -0,0 +1,546 @@
+#if defined(LOKI_ENABLE_INTEGRATION_TEST_HOOKS)
+    const time_t TIME_BETWEEN_RELAY = 0;
+#else
+    const time_t TIME_BETWEEN_RELAY = 60 * 2;
+#endif
+
+    time_t const now = time(nullptr);
+    std::vector<quorum_vote_t> result;
+    for (deregister_pool_entry const &pool_entry : m_deregister_pool)
+    {
+      for (pool_vote_entry const &vote_entry : pool_entry.votes)
+      {
+        const time_t last_sent = now - vote_entry.time_last_sent_p2p;
+        if (last_sent > TIME_BETWEEN_RELAY)
+          result.push_back(vote_entry.vote);
+      }
+    }
+
+    for (checkpoint_pool_entry const &pool_entry : m_checkpoint_pool)
+    {
+      for (pool_vote_entry const &vote_entry : pool_entry.votes)
+      {
+        const time_t last_sent = now - vote_entry.time_last_sent_p2p;
+        if (last_sent > TIME_BETWEEN_RELAY)
+          result.push_back(vote_entry.vote);
+      }
+    }
+
+    return result;
+  }
+
+  // return: True if the vote was unique
+  template <typename T>
+  static bool add_vote_to_pool_if_unique(T &vote_pool, quorum_vote_t const &vote)
+  {
+    auto vote_it = std::find_if(vote_pool.votes.begin(), vote_pool.votes.end(), [&vote](pool_vote_entry const &pool_entry) {
+        assert(pool_entry.vote.group == vote.group);
+        return (pool_entry.vote.index_in_group == vote.index_in_group);
+    });
+
+    bool result = false;
+    if (vote_it == vote_pool.votes.end())
+    {
+      pool_vote_entry entry = {};
+      entry.vote            = vote;
+      vote_pool.votes.push_back(entry);
+      result                = true;
+    }
+
+    return result;
+  }
+
+  std::vector<pool_vote_entry> voting_pool::add_pool_vote_if_unique(uint64_t latest_height,
+                                                                    const quorum_vote_t& vote,
+                                                                    cryptonote::vote_verification_context& vvc,
+                                                                    const service_nodes::testing_quorum &quorum)
+  {
+    std::vector<pool_vote_entry> result = {};
+
+    if (!verify_vote(vote, latest_height, vvc, quorum))
+    {
+      LOG_PRINT_L1("Signature verification failed for deregister vote");
+      return result;
+    }
+
+    CRITICAL_REGION_LOCAL(m_lock);
+    switch(vote.type)
+    {
+      default:
+      {
+        LOG_PRINT_L1("Unhandled vote type with value: " << (int)vote.type);
+        assert("Unhandled vote type" == 0);
+        return result;
+      };
+
+      case quorum_type::uptime_deregister:
+      {
+        time_t const now = time(NULL);
+        auto it = std::find_if(m_deregister_pool.begin(), m_deregister_pool.end(), [&vote](deregister_pool_entry const &entry) {
+            return (entry.height == vote.block_height && entry.worker_index == vote.deregister.worker_index);
+        });
+
+        if (it == m_deregister_pool.end())
+        {
```

Comment:
add a constructor to `deregister_pool_entry`, and this could be:

```c++
it = m_deregister_pool.emplace(m_deregister_pool.end(), vote.block_height, vote.deregister.worker_index);
```

Reply:
Added